File: packages/react-native-hms/ios/HMSHLSPlayerManager.m

#import "React/RCTViewManager.h"
#import <Foundation/Foundation.h>

// Under the New Architecture, `<HMSHLSPlayer />` is rendered via the
// Fabric component view in HMSHLSPlayerComponentView.mm. The legacy
// paper RCT_EXTERN_MODULE + RCT_EXPORT_VIEW_PROPERTY + RCT_EXTERN_METHOD
// declarations below are only used in old-arch / interop mode, so we
// guard the whole block with #if !RCT_NEW_ARCH_ENABLED to prevent
// double-registration.
#if !RCT_NEW_ARCH_ENABLED

@interface RCT_EXTERN_MODULE(HMSHLSPlayerManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(url, NSString);
RCT_EXPORT_VIEW_PROPERTY(enableStats, BOOL);
RCT_EXPORT_VIEW_PROPERTY(enableControls, BOOL);
RCT_EXPORT_VIEW_PROPERTY(onDataReturned, RCTDirectEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onHmsHlsPlaybackEvent, RCTDirectEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onHmsHlsStatsEvent, RCTDirectEventBlock);

RCT_EXTERN_METHOD(play:(nonnull NSNumber *)node url:(nullable NSString *)url)
RCT_EXTERN_METHOD(stop:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(pause:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(resume:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(seekToLivePosition:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(seekForward:(nonnull NSNumber *)node seconds:(nonnull NSNumber *)seconds)
RCT_EXTERN_METHOD(seekBackward:(nonnull NSNumber *)node seconds:(nonnull NSNumber *)seconds)
RCT_EXTERN_METHOD(setVolume:(nonnull NSNumber *)node level:(nonnull NSNumber *)level)
RCT_EXTERN_METHOD(areClosedCaptionSupported:(nonnull NSNumber *)node requestId:(nonnull NSNumber *)requestId)
RCT_EXTERN_METHOD(isClosedCaptionEnabled:(nonnull NSNumber *)node requestId:(nonnull NSNumber *)requestId)
RCT_EXTERN_METHOD(enableClosedCaption:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(disableClosedCaption:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(getPlayerDurationDetails:(nonnull NSNumber *)node requestId:(nonnull NSNumber *)requestId)

@end

#endif  // !RCT_NEW_ARCH_ENABLED

File: packages/react-native-hms/android/src/newarch/java/com/reactnativehmssdk/HMSHLSPlayerManager.kt
package com.reactnativehmssdk

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.viewmanagers.HMSHLSPlayerManagerDelegate
import com.facebook.react.viewmanagers.HMSHLSPlayerManagerInterface

/**
 * HMSHLSPlayerManager — new-arch (Fabric) view manager for `<HMSHLSPlayer />`.
 *
 * Phase 1 / 1C-4 of the New Architecture migration. Compiled only when
 * the consumer's app uses new architecture.
 *
 * Differences from the old-arch wrapper:
 *   - Implements the Codegen-generated `HMSHLSPlayerManagerInterface<HMSHLSPlayer>`
 *     (typed prop setters AND typed command methods that Fabric calls directly).
 *   - Uses `HMSHLSPlayerManagerDelegate` for prop dispatch via Fabric's
 *     mounting mechanism.
 *   - Drops @ReactProp annotations — Fabric doesn't scan for them.
 *   - Implements receiveCommand(commandName: String, args) for Fabric's
 *     string-based command dispatch (in addition to numeric Int dispatch
 *     via the typed command methods).
 *
 * All logic delegates to HMSHLSPlayerManagerImpl in src/main/.
 */
class HMSHLSPlayerManager :
  SimpleViewManager<HMSHLSPlayer>(),
  HMSHLSPlayerManagerInterface<HMSHLSPlayer> {
  private val mDelegate: ViewManagerDelegate<HMSHLSPlayer> = HMSHLSPlayerManagerDelegate(this)

  override fun getName(): String = HMSHLSPlayerManagerImpl.REACT_CLASS

  override fun getDelegate(): ViewManagerDelegate<HMSHLSPlayer> = mDelegate

  override fun createViewInstance(reactContext: ThemedReactContext): HMSHLSPlayer =
    HMSHLSPlayerManagerImpl.createViewInstance(reactContext)

  override fun onDropViewInstance(view: HMSHLSPlayer) {
    super.onDropViewInstance(view)
    HMSHLSPlayerManagerImpl.onDropViewInstance(view)
  }

  override fun getExportedCustomDirectEventTypeConstants(): Map<String, Any> =
    HMSHLSPlayerManagerImpl.getExportedCustomDirectEventTypeConstants()

  /** Fabric command dispatch by string name (Codegen routes here via the delegate). */
  override fun receiveCommand(
    root: HMSHLSPlayer,
    commandId: String?,
    args: ReadableArray?,
  ) {
    if (commandId != null) {
      HMSHLSPlayerManagerImpl.receiveCommandByName(root, commandId, args)
    }
  }

  // ─────────────────────────────────────────────────────────────────
  // Codegen-interface methods — typed prop setters
  // ─────────────────────────────────────────────────────────────────

  override fun setUrl(
    view: HMSHLSPlayer,
    value: String?,
  ) = HMSHLSPlayerManagerImpl.setUrl(view, value)

  override fun setEnableStats(
    view: HMSHLSPlayer,
    value: Boolean,
  ) = HMSHLSPlayerManagerImpl.setEnableStats(view, value)

  override fun setEnableControls(
    view: HMSHLSPlayer,
    value: Boolean,
  ) = HMSHLSPlayerManagerImpl.setEnableControls(view, value)

  // ─────────────────────────────────────────────────────────────────
  // Codegen-interface methods — typed imperative commands (13)
  // ─────────────────────────────────────────────────────────────────

  override fun play(
    view: HMSHLSPlayer,
    url: String?,
  ) = HMSHLSPlayerManagerImpl.play(view, url)

  override fun stop(view: HMSHLSPlayer) = HMSHLSPlayerManagerImpl.stop(view)

  override fun pause(view: HMSHLSPlayer) = HMSHLSPlayerManagerImpl.pause(view)

  override fun resume(view: HMSHLSPlayer) = HMSHLSPlayerManagerImpl.resume(view)

  override fun seekToLivePosition(view: HMSHLSPlayer) = HMSHLSPlayerManagerImpl.seekToLivePosition(view)

  override fun seekForward(
    view: HMSHLSPlayer,
    seconds: Double,
  ) = HMSHLSPlayerManagerImpl.seekForward(view, seconds)

  override fun seekBackward(
    view: HMSHLSPlayer,
    seconds: Double,
  ) = HMSHLSPlayerManagerImpl.seekBackward(view, seconds)

  override fun setVolume(
    view: HMSHLSPlayer,
    level: Int,
  ) = HMSHLSPlayerManagerImpl.setVolume(view, level)

  override fun areClosedCaptionSupported(
    view: HMSHLSPlayer,
    requestId: Int,
  ) = HMSHLSPlayerManagerImpl.areClosedCaptionSupported(view, requestId)

  override fun isClosedCaptionEnabled(
    view: HMSHLSPlayer,
    requestId: Int,
  ) = HMSHLSPlayerManagerImpl.isClosedCaptionEnabled(view, requestId)

  override fun enableClosedCaption(view: HMSHLSPlayer) = HMSHLSPlayerManagerImpl.enableClosedCaption(view)

  override fun disableClosedCaption(view: HMSHLSPlayer) = HMSHLSPlayerManagerImpl.disableClosedCaption(view)

  override fun getPlayerDurationDetails(
    view: HMSHLSPlayer,
    requestId: Int,
  ) = HMSHLSPlayerManagerImpl.getPlayerDurationDetails(view, requestId)
}
